ingest/rules/group_segments.smk

```diff
@@ -0,0 +1,41 @@
+
+rule group_segments:
+    input:
+        metadata="data/metadata_merged.tsv",
+        resolutions=config["grouping"]["resolutions"],
+    output:
+        metadata="results/metadata.tsv"
+    params:
+        common_strain_fields = config["grouping"]["common_strain_fields"],
+        segments = segments,
+    shell:
+        r"""
+        python3 scripts/group_segments.py \
+            --metadata {input.metadata} \
+            --common-strain-fields {params.common_strain_fields} \
+            --segments {params.segments} \
+            --resolutions {input.resolutions} \
+            --output-metadata {output.metadata}
+        """
+
+rule subset_sequences_by_segment:
+    input:
+        metadata = "results/metadata.tsv",
+        sequences = "data/sequences.fasta",
+    output:
+        kv_map = temp("data/kv-map_{segment}.tsv"),
+        sequences = "results/{segment}/sequences.fasta",
+    params:
+        columns = lambda w: f"accession_{w.segment},strain"
+    shell:
+        r"""
+        cat results/metadata.tsv \
+            | csvtk cut -t -f {params.columns} \
+            | grep -v '^\t' \
+            > {output.kv_map} && \
+        seqkit replace \
+            -p "(.*)" --replacement "{{kv}}" --kv-file {output.kv_map} \
+            {input.sequences} \
+            | seqkit grep -v -r -p '^\t' \
+            > {output.sequences}
```

Comment:
The `grep` and `seqkit grep` commands are not working as expected here, all three `results/{segment}/sequences.fasta` files still have all 2189 records with various empty strain names.  

My quick and dirty patch to filter the sequences: 
```diff
diff --git a/ingest/rules/group_segments.smk b/ingest/rules/group_segments.smk
index e901f41..7411dd1 100644
--- a/ingest/rules/group_segments.smk
+++ b/ingest/rules/group_segments.smk
@@ -46,16 +46,19 @@ rule subset_sequences_by_segment:
         kv_map = temp("data/kv-map_{segment}.tsv"),
         sequences = "results/{segment}/sequences.fasta",
     params:
-        columns = lambda w: f"accession_{w.segment},strain"
+        columns = lambda w: f"accession_{w.segment},strain",
+        filter_exp = lambda w: f"len($accession_{w.segment})>0",
+        drop_key = "__DROP_SEQ__",
     shell:
         r"""
         cat results/metadata.tsv \
             | csvtk cut -t -f {params.columns} \
-            | grep -v '^\t' \
+            | csvtk filter2 -t -f {params.filter_exp:q} \
+            | csvtk del-header -t \
             > {output.kv_map} && \
         seqkit replace \
-            -p "(.*)" --replacement "{{kv}}" --kv-file {output.kv_map} \
+            -p "(.*)" --replacement "{{kv}}" --kv-file {output.kv_map} -m {params.drop_key:q} \
             {input.sequences} \
-            | seqkit grep -v -r -p '^\t' \
+            | seqkit grep -v -r -p '^{params.drop_key:q}$' \
             > {output.sequences}
         """
```

Reply:
Oh. Interesting. Not for me!

```console
for i in results/*/sequences.fasta; do echo "$i $(grep '>' $i | wc -l)"; done
results/L/sequences.fasta      680
results/M/sequences.fasta      679
results/S/sequences.fasta      792
```

I'm on seqkit Version: 2.8.0, csvtk v0.30.0, grep (BSD grep, GNU compatible) 2.6.0-FreeBSD.

Anyhow, I've changed the rule to be (essentially) what you suggested above -- would you mind trying it out? It'd be interesting to know _why_ it didn't work for you, but if it didn't work for you it undoubtedly won't work for someone else either.

Reply:
The BSD grep that ships with macOS interprets `\t` and GNU grep does not. Compare:

```console
tom@bloop ~ $ grep --version
grep (BSD grep, GNU compatible) 2.6.0-FreeBSD
tom@bloop ~ $ printf '%s\t%s\n' a b "" d e f | grep -v '^\t'
a	b
e	f
```

```console
tom@clack ~ $ grep --version
grep (GNU grep) 3.11
tom@clack ~ $ printf '%s\t%s\n' a b "" d e f | grep -v '^\t'
a	b
	d
e	f
```

Note that our runtimes all include GNU coreutils, but *not* GNU grep. They should probably standardize on GNU grep! This isn't the only significant difference in behaviour between the grep variants. While we're at it, we should standardize on GNU awk (gawk) and GNU sed too if we're not already.

Reply:
https://github.com/nextstrain/conda-base/pull/94

Reply:
Ah, I'm running the workflow with the Docker runtime:
```
$ seqkit version && csvtk version && grep --version
seqkit v2.2.0
csvtk v0.30.0
grep (GNU grep) 3.8
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Mike Haertel and others; see
<https://git.sv.gnu.org/cgit/grep.git/tree/AUTHORS>.
```

Reply:
The BSD vs GNU grep is part of the story, but `seqkit grep` also differs between (my) MacOS version (Version: 2.8.0) and the latest one in our docker container (Version: 2.2.0). Unsure if that's OS-specific or version-specific.

There's no difference in the `seqkit replace` command output, however `seqkit grep -v -r -p '^\t'` behaves differently - it works as intended¹ on MacOS but not Docker (where sequences with empty strain names remain).

I'll keep using the explicit `__DROP__` for this PR.

¹ Not so sure what's going on here -- the sequence names we want to drop (after `seqkit replace`) have empty FASTA headers (i.e. `>\n`) so I don't know why `^\t` would work. But it does. Similarly `^$` works on MacOS but not Docker.

Reply:
(For the purposes of this PR, the new code behaves the same across an ambient MacOS runtime and our Docker runtime, so I consider this resolved but will leave open in case we continue the discussion)

Reply:
`seqkit grep` changed its behaviour with regard to the interaction of `-v` and empty ids/names sometime between 2.2.0 and 2.8.2 (probably earlier). Consider:

```console
$ seqkit version
seqkit v2.8.2

$ printf '>%s\n%s\n' foo ATCG "" CTAG bar XYZ
>foo
ATCG
>
CTAG
>bar
XYZ

$ printf '>%s\n%s\n' foo ATCG "" CTAG bar XYZ | seqkit grep -v -p 'fjklsdf'
>foo
ATCG
>bar
XYZ
```

and

```console
$ seqkit version
seqkit v2.2.0

$ printf '>%s\n%s\n' foo ATCG "" CTAG bar XYZ | seqkit grep -v -p 'fjklsdf'
>foo
ATCG
>
CTAG

>bar
XYZ
```

In 2.2.0, it seems that an empty id/name _always_ matches under `-v`. In 2.8.2, an empty id/name _never_ matches under `-v`.